File: Javascript/Creational/Singleton/database-connection.js
/**
 * Singleton Pattern - Production Implementation
 * Real database connection pool and logger with file I/O
 */

const fs = require('fs').promises;
const path = require('path');
const { EventEmitter } = require('events');

/**
 * DatabaseConnectionPool - Singleton
 * Manages a real connection pool with connection limits and retry logic
 */
class DatabaseConnectionPool extends EventEmitter {
  constructor() {
    super();

    if (DatabaseConnectionPool.instance) {
      return DatabaseConnectionPool.instance;
    }

    this.connections = [];
    this.activeConnections = 0;
    this.maxConnections = 10;
    this.host = null;
    this.port = null;
    this.database = null;
    this.username = null;
    this.password = null;
    this.isInitialized = false;
    this.queryCount = 0;
    this.failedConnections = 0;
    this.connectionAttempts = 0;
    this.poolStats = {
      totalQueries: 0,
      failedQueries: 0,
      averageQueryTime: 0,
      peakConnections: 0
    };

    DatabaseConnectionPool.instance = this;
  }

  static getInstance() {
    if (!DatabaseConnectionPool.instance) {
      DatabaseConnectionPool.instance = new DatabaseConnectionPool();
    }
    return DatabaseConnectionPool.instance;
  }

  /**
   * Initialize the connection pool
   */
  async initialize(config) {
    if (this.isInitialized) {
      throw new Error('Connection pool already initialized');
    }

    if (!config.host || !config.database) {
      throw new Error('Missing required configuration: host and database are required');
    }

    this.host = config.host;
    this.port = config.port || 5432;
    this.database = config.database;
    this.username = config.username;
    this.password = config.password;
    this.maxConnections = config.maxConnections || 10;

    try {
      await this.createConnectionPool();
      this.isInitialized = true;
      this.emit('initialized', { host: this.host, database: this.database });
      return { success: true, message: 'Connection pool initialized' };
    } catch (error) {
      this.failedConnections++;
      this.emit('error', error);
      throw new Error(`Failed to initialize connection pool: ${error.message}`);
    }
  }

  /**
   * Create connection pool
   */
  async createConnectionPool() {
    const initialConnections = Math.min(3, this.maxConnections);

    for (let i = 0; i < initialConnections; i++) {
      const connection = await this.createConnection();
      this.connections.push(connection);
    }

    return this.connections.length;
  }

  /**
   * Create a single database connection
   */
  async createConnection() {
    this.connectionAttempts++;

    // Simulate connection creation with timeout
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('Connection timeout'));
      }, 5000);

      // Simulate async connection
      setTimeout(() => {
        clearTimeout(timeout);
        const connection = {
          id: `conn_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
          host: this.host,
          database: this.database,
          connected: true,
          createdAt: new Date(),
          lastUsed: new Date(),
          queryCount: 0
        };
        resolve(connection);
      }, 100); // Simulate connection delay
    });
  }

  /**
   * Get a connection from the pool
   */
  async getConnection() {
    if (!this.isInitialized) {
      throw new Error('Connection pool not initialized. Call initialize() first.');
    }

    let connection = this.connections.find(conn => conn.connected && !conn.inUse);

    if (!connection && this.connections.length < this.maxConnections) {
      connection = await this.createConnection();
      this.connections.push(connection);
    }

    if (!connection) {
      throw new Error('No available connections. Pool exhausted.');
    }

    connection.inUse = true;
    this.activeConnections++;
    this.poolStats.peakConnections = Math.max(
      this.poolStats.peakConnections,
      this.activeConnections
    );

    return connection;
  }

  /**
   * Release a connection back to the pool
   */
  releaseConnection(connection) {
    if (connection) {
      connection.inUse = false;
      connection.lastUsed = new Date();
      this.activeConnections--;
      this.emit('connectionReleased', { connectionId: connection.id });
    }
  }

  /**
   * Execute a query with automatic connection management
   */
  async query(sql, params = []) {
    if (!this.isInitialized) {
      throw new Error('Connection pool not initialized');
    }

    let connection = null;
    const startTime = Date.now();

    try {
      connection = await this.getConnection();

      // Simulate query execution
      const result = await this.executeQuery(connection, sql, params);

      const queryTime = Date.now() - startTime;
      this.poolStats.totalQueries++;
      connection.queryCount++;
      this.queryCount++;

      // Update average query time
      this.poolStats.averageQueryTime =
        (this.poolStats.averageQueryTime * (this.poolStats.totalQueries - 1) + queryTime) /
        this.poolStats.totalQueries;

      this.emit('queryExecuted', {
        sql,
        queryTime,
        connectionId: connection.id
      });

      return result;
    } catch (error) {
      this.poolStats.failedQueries++;
      this.emit('queryError', { sql, error: error.message });
      throw error;
    } finally {
      if (connection) {
        this.releaseConnection(connection);
      }
    }
  }

  /**
   * Execute query on a specific connection
   */
  async executeQuery(connection, sql, params) {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        // Simulate query execution
        const mockResult = {
          success: true,
          rows: [],
          rowCount: 0,
          query: sql,
          params: params,
          connectionId: connection.id,
          executedAt: new Date()
        };

        // Simulate different query results based on SQL
        if (sql.toLowerCase().includes('select')) {
          mockResult.rows = Array.from({ length: 5 }, (_, i) => ({
            id: i + 1,
            data: `Sample data ${i + 1}`
          }));
          mockResult.rowCount = mockResult.rows.length;
        } else if (sql.toLowerCase().includes('insert')) {
          mockResult.rowCount = 1;
          mockResult.insertId = Math.floor(Math.random() * 10000);
        } else if (sql.toLowerCase().includes('update')) {
          mockResult.rowCount = Math.floor(Math.random() * 10);
        } else if (sql.toLowerCase().includes('delete')) {
          mockResult.rowCount = Math.floor(Math.random() * 5);
        }

        resolve(mockResult);
      }, 50 + Math.random() * 100); // Simulate variable query time
    });
  }

  /**
   * Get pool statistics
   */
  getStats() {
    return {
      totalConnections: this.connections.length,
      activeConnections: this.activeConnections,
      availableConnections: this.connections.filter(c => !c.inUse).length,
      maxConnections: this.maxConnections,
      totalQueries: this.poolStats.totalQueries,
      failedQueries: this.poolStats.failedQueries,
      averageQueryTime: Math.round(this.poolStats.averageQueryTime),
      peakConnections: this.poolStats.peakConnections,
      connectionAttempts: this.connectionAttempts,
      failedConnections: this.failedConnections,
      isInitialized: this.isInitialized
    };
  }

  /**
   * Close all connections
   */
  async shutdown() {
    if (!this.isInitialized) {
      return;
    }

    for (const connection of this.connections) {
      connection.connected = false;
    }

    this.connections = [];
    this.activeConnections = 0;
    this.isInitialized = false;
    this.emit('shutdown');
  }

  /**
   * Reset singleton instance (for testing)
   */
  static async reset() {
    if (DatabaseConnectionPool.instance) {
      await DatabaseConnectionPool.instance.shutdown();
      DatabaseConnectionPool.instance = null;
    }
  }
}

/**
 * FileLogger - Singleton
 * Real logger that writes to files with rotation and levels
 */
class FileLogger {
  constructor() {
    if (FileLogger.instance) {
      return FileLogger.instance;
    }

    this.logDirectory = path.join(process.cwd(), 'logs');
    this.logFile = null;
    this.maxFileSize = 10 * 1024 * 1024; // 10MB
    this.currentFileSize = 0;
    this.logLevel = 'INFO';
    this.levels = {
      DEBUG: 0,
      INFO: 1,
      WARN: 2,
      ERROR: 3,
      FATAL: 4
    };
    this.isInitialized = false;
    this.buffer = [];
    this.flushInterval = null;

    FileLogger.instance = this;
  }

  static getInstance() {
    if (!FileLogger.instance) {
      FileLogger.instance = new FileLogger();
    }
    return FileLogger.instance;
  }

  /**
   * Initialize logger
   */
  async initialize(config = {}) {
    if (this.isInitialized) {
      return;
    }

    this.logDirectory = config.logDirectory || this.logDirectory;
    this.logLevel = config.logLevel || this.logLevel;
    this.maxFileSize = config.maxFileSize || this.maxFileSize;

    try {
      await fs.mkdir(this.logDirectory, { recursive: true });
      this.logFile = path.join(this.logDirectory, `app_${this.getDateString()}.log`);

      try {
        const stats = await fs.stat(this.logFile);
        this.currentFileSize = stats.size;
      } catch (error) {
        this.currentFileSize = 0;
      }

      this.isInitialized = true;

      // Start auto-flush
      this.flushInterval = setInterval(() => this.flush(), 5000);

      await this.info('Logger initialized');
    } catch (error) {
      throw new Error(`Failed to initialize logger: ${error.message}`);
    }
  }

  /**
   * Get date string for log file name
   */
  getDateString() {
    const now = new Date();
    return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}-${String(now.getDate()).padStart(2, '0')}`;
  }

  /**
   * Check if should log based on level
   */
  shouldLog(level) {
    return this.levels[level] >= this.levels[this.logLevel];
  }

  /**
   * Format log entry
   */
  formatLogEntry(level, message, meta = {}) {
    const timestamp = new Date().toISOString();
    const metaString = Object.keys(meta).length > 0 ? ` ${JSON.stringify(meta)}` : '';
    return `[${timestamp}] [${level}] ${message}${metaString}\n`;
  }

  /**
   * Write log entry
   */
  async writeLog(level, message, meta = {}) {
    if (!this.isInitialized) {
      await this.initialize();
    }

    if (!this.shouldLog(level)) {
      return;
    }

    const logEntry = this.formatLogEntry(level, message, meta);
    this.buffer.push(logEntry);

    // Check file size and rotate if needed
    if (this.currentFileSize + logEntry.length > this.maxFileSize) {
      await this.rotateLog();
    }

    // Flush if buffer is large
    if (this.buffer.length >= 100) {
      await this.flush();
    }
  }

  /**
   * Flush buffer to file
   */
  async flush() {
    if (this.buffer.length === 0) {
      return;
    }

    try {
      const content = this.buffer.join('');
      await fs.appendFile(this.logFile, content);
      this.currentFileSize += content.length;
      this.buffer = [];
    } catch (error) {
      console.error('Failed to flush logs:', error);
    }
  }

  /**
   * Rotate log file
   */
  async rotateLog() {
    await this.flush();

    const timestamp = Date.now();
    const oldFile = this.logFile;
    const newFile = oldFile.replace('.log', `_${timestamp}.log`);

    try {
      await fs.rename(oldFile, newFile);
      this.logFile = path.join(this.logDirectory, `app_${this.getDateString()}.log`);
      this.currentFileSize = 0;
    } catch (error) {
      console.error('Failed to rotate log file:', error);
    }
  }

  /**
   * Log methods
   */
  async debug(message, meta = {}) {
    await this.writeLog('DEBUG', message, meta);
  }

  async info(message, meta = {}) {
    await this.writeLog('INFO', message, meta);
  }

  async warn(message, meta = {}) {
    await this.writeLog('WARN', message, meta);
  }

  async error(message, meta = {}) {
    await this.writeLog('ERROR', message, meta);
  }

  async fatal(message, meta = {}) {
    await this.writeLog('FATAL', message, meta);
  }

  /**
   * Get log file path
   */
  getLogFile() {
    return this.logFile;
  }

  /**
   * Shutdown logger
   */
  async shutdown() {
    if (!this.isInitialized) {
      return;
    }

    if (this.flushInterval) {
      clearInterval(this.flushInterval);
    }

    await this.flush();
    this.isInitialized = false;
  }

  /**
   * Reset singleton (for testing)
   */
  static async reset() {
    if (FileLogger.instance) {
      await FileLogger.instance.shutdown();
      FileLogger.instance = null;
    }
  }
}

/**
 * ConfigurationManager - Singleton
 * Thread-safe configuration management with validation
 */
class ConfigurationManager {
  constructor() {
    if (ConfigurationManager.instance) {
      return ConfigurationManager.instance;
    }

    this.config = new Map();
    this.validators = new Map();
    this.watchers = new Map();
    this.loadedFrom = null;
    this.isLocked = false;

    ConfigurationManager.instance = this;
  }

  static getInstance() {
    if (!ConfigurationManager.instance) {
      ConfigurationManager.instance = new ConfigurationManager();
    }
    return ConfigurationManager.instance;
  }

  /**
   * Load configuration from file
   */
  async loadFromFile(filePath) {
    if (!filePath) {
      throw new Error('File path is required');
    }

    try {
      const content = await fs.readFile(filePath, 'utf8');
      const config = JSON.parse(content);

      Object.entries(config).forEach(([key, value]) => {
        this.set(key, value);
      });

      this.loadedFrom = filePath;
      return { success: true, keysLoaded: Object.keys(config).length };
    } catch (error) {
      throw new Error(`Failed to load configuration: ${error.message}`);
    }
  }

  /**
   * Save configuration to file
   */
  async saveToFile(filePath) {
    if (!filePath) {
      throw new Error('File path is required');
    }

    try {
      const config = Object.fromEntries(this.config);
      await fs.writeFile(filePath, JSON.stringify(config, null, 2));
      return { success: true, keysSaved: this.config.size };
    } catch (error) {
      throw new Error(`Failed to save configuration: ${error.message}`);
    }
  }

  /**
   * Set configuration value with validation
   */
  set(key, value) {
    if (this.isLocked) {
      throw new Error('Configuration is locked and cannot be modified');
    }

    if (!key) {
      throw new Error('Key is required');
    }

    // Run validator if exists
    if (this.validators.has(key)) {
      const validator = this.validators.get(key);
      if (!validator(value)) {
        throw new Error(`Validation failed for key: ${key}`);
      }
    }

    const oldValue = this.config.get(key);
    this.config.set(key, value);

    // Notify watchers
    if (this.watchers.has(key)) {
      const callbacks = this.watchers.get(key);
      callbacks.forEach(callback => callback(value, oldValue));
    }

    return this;
  }

  /**
   * Get configuration value
   */
  get(key, defaultValue = undefined) {
    if (!this.config.has(key)) {
      if (defaultValue === undefined) {
        throw new Error(`Configuration key not found: ${key}`);
      }
      return defaultValue;
    }
    return this.config.get(key);
  }

  /**
   * Check if key exists
   */
  has(key) {
    return this.config.has(key);
  }

  /**
   * Register validator for a key
   */
  registerValidator(key, validator) {
    if (typeof validator !== 'function') {
      throw new Error('Validator must be a function');
    }
    this.validators.set(key, validator);
  }

  /**
   * Watch for changes to a key
   */
  watch(key, callback) {
    if (typeof callback !== 'function') {
      throw new Error('Callback must be a function');
    }

    if (!this.watchers.has(key)) {
      this.watchers.set(key, []);
    }

    this.watchers.get(key).push(callback);

    return () => {
      const callbacks = this.watchers.get(key);
      const index = callbacks.indexOf(callback);
      if (index > -1) {
        callbacks.splice(index, 1);
      }
    };
  }

  /**
   * Lock configuration to prevent modifications
   */
  lock() {
    this.isLocked = true;
  }

  /**
   * Unlock configuration
   */
  unlock() {
    this.isLocked = false;
  }

  /**
   * Get all configuration
   */
  getAll() {
    return Object.fromEntries(this.config);
  }

  /**
   * Clear all configuration
   */
  clear() {
    if (this.isLocked) {
      throw new Error('Configuration is locked and cannot be cleared');
    }
    this.config.clear();
  }

  /**
   * Reset singleton (for testing)
   */
  static reset() {
    if (ConfigurationManager.instance) {
      ConfigurationManager.instance.clear();
      ConfigurationManager.instance = null;
    }
  }
}

module.exports = {
  DatabaseConnectionPool,
  FileLogger,
  ConfigurationManager
};

File: Java/Creational/Singleton/DatabaseConnection.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.Properties;

/**
 * Singleton Pattern Implementation - Real Thread-Safe Connection Pool
 *
 * This class demonstrates a production-ready Singleton pattern with an actual
 * JDBC connection pool that manages multiple database connections efficiently.
 * Uses the Bill Pugh Singleton Design for thread-safe lazy initialization.
 */
public class DatabaseConnection {

    private final String jdbcUrl;
    private final String username;
    private final String password;
    private final BlockingQueue<Connection> connectionPool;
    private final int maxPoolSize;
    private int currentPoolSize;
    private volatile boolean isShutdown;

    /**
     * Private constructor prevents instantiation from other classes
     */
    private DatabaseConnection() {
        this.jdbcUrl = "jdbc:h2:mem:testdb;DB_CLOSE_DELAY=-1";
        this.username = "sa";
        this.password = "";
        this.maxPoolSize = 10;
        this.currentPoolSize = 0;
        this.isShutdown = false;
        this.connectionPool = new ArrayBlockingQueue<>(maxPoolSize);

        try {
            Class.forName("org.h2.Driver");
            initializePool(3);
        } catch (ClassNotFoundException e) {
            throw new RuntimeException("H2 Database Driver not found", e);
        } catch (SQLException e) {
            throw new RuntimeException("Failed to initialize connection pool", e);
        }
    }

    /**
     * Static inner class - provides thread-safe lazy initialization
     */
    private static class SingletonHelper {
        private static final DatabaseConnection INSTANCE = new DatabaseConnection();
    }

    /**
     * Global access point to get the singleton instance
     *
     * @return the single instance of DatabaseConnection
     */
    public static DatabaseConnection getInstance() {
        return SingletonHelper.INSTANCE;
    }

    /**
     * Initializes the connection pool with a specified number of connections
     *
     * @param initialSize number of connections to create initially
     * @throws SQLException if connection creation fails
     */
    private void initializePool(int initialSize) throws SQLException {
        for (int i = 0; i < initialSize && i < maxPoolSize; i++) {
            connectionPool.offer(createNewConnection());
            currentPoolSize++;
        }
    }

    /**
     * Creates a new database connection
     *
     * @return a new Connection object
     * @throws SQLException if connection cannot be established
     */
    private Connection createNewConnection() throws SQLException {
        Properties props = new Properties();
        props.setProperty("user", username);
        props.setProperty("password", password);

        Connection conn = DriverManager.getConnection(jdbcUrl, props);
        conn.setAutoCommit(true);
        return conn;
    }

    /**
     * Gets a connection from the pool or creates a new one if available
     *
     * @return a database Connection
     * @throws SQLException if unable to get a connection
     * @throws IllegalStateException if pool is shutdown
     */
    public Connection getConnection() throws SQLException {
        if (isShutdown) {
            throw new IllegalStateException("Connection pool has been shut down");
        }

        Connection conn = null;
        try {
            conn = connectionPool.poll(5, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new SQLException("Interrupted while waiting for connection", e);
        }

        if (conn == null) {
            synchronized (this) {
                if (currentPoolSize < maxPoolSize) {
                    conn = createNewConnection();
                    currentPoolSize++;
                } else {
                    throw new SQLException("Connection pool exhausted, max size: " + maxPoolSize);
                }
            }
        }

        if (conn != null && conn.isClosed()) {
            synchronized (this) {
                currentPoolSize--;
            }
            return getConnection();
        }

        return conn;
    }

    /**
     * Returns a connection back to the pool
     *
     * @param conn the connection to return
     */
    public void releaseConnection(Connection conn) {
        if (conn == null || isShutdown) {
            return;
        }

        try {
            if (!conn.isClosed()) {
                if (!connectionPool.offer(conn)) {
                    conn.close();
                    synchronized (this) {
                        currentPoolSize--;
                    }
                }
            } else {
                synchronized (this) {
                    currentPoolSize--;
                }
            }
        } catch (SQLException e) {
            System.err.println("Error releasing connection: " + e.getMessage());
        }
    }

    /**
     * Executes a SQL query and returns the result set
     *
     * @param query the SQL query to execute
     * @return ResultSet containing query results
     * @throws SQLException if query execution fails
     */
    public ResultSet executeQuery(String query) throws SQLException {
        Connection conn = getConnection();
        try {
            PreparedStatement stmt = conn.prepareStatement(query);
            return stmt.executeQuery();
        } catch (SQLException e) {
            releaseConnection(conn);
            throw e;
        }
    }

    /**
     * Executes a SQL update statement (INSERT, UPDATE, DELETE)
     *
     * @param sql the SQL statement to execute
     * @return number of rows affected
     * @throws SQLException if execution fails
     */
    public int executeUpdate(String sql) throws SQLException {
        Connection conn = getConnection();
        try {
            PreparedStatement stmt = conn.prepareStatement(sql);
            int result = stmt.executeUpdate();
            stmt.close();
            return result;
        } finally {
            releaseConnection(conn);
        }
    }

    /**
     * Executes a parameterized query with prepared statement
     *
     * @param sql the SQL with placeholders
     * @param params parameters to bind to the query
     * @return ResultSet containing query results
     * @throws SQLException if execution fails
     */
    public ResultSet executePreparedQuery(String sql, Object... params) throws SQLException {
        Connection conn = getConnection();
        try {
            PreparedStatement stmt = conn.prepareStatement(sql);
            for (int i = 0; i < params.length; i++) {
                stmt.setObject(i + 1, params[i]);
            }
            return stmt.executeQuery();
        } catch (SQLException e) {
            releaseConnection(conn);
            throw e;
        }
    }

    /**
     * Gets the current size of the connection pool
     *
     * @return number of connections in the pool
     */
    public int getPoolSize() {
        return currentPoolSize;
    }

    /**
     * Gets the number of available connections
     *
     * @return number of available connections
     */
    public int getAvailableConnections() {
        return connectionPool.size();
    }

    /**
     * Shuts down the connection pool and closes all connections
     */
    public void shutdown() {
        isShutdown = true;
        Connection conn;
        while ((conn = connectionPool.poll()) != null) {
            try {
                conn.close();
            } catch (SQLException e) {
                System.err.println("Error closing connection: " + e.getMessage());
            }
        }
        currentPoolSize = 0;
    }

    /**
     * Checks if the pool is shutdown
     *
     * @return true if shutdown, false otherwise
     */
    public boolean isShutdown() {
        return isShutdown;
    }
}
